File: device-discovery/device_discovery/policy/job.py
#!/usr/bin/env python
# Copyright 2024 NetBox Labs Inc
"""Device Discovery Job Tracking."""

import threading
import uuid
from dataclasses import dataclass, field
from datetime import datetime
from enum import Enum


class JobStatus(str, Enum):
    """Job status enumeration."""

    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"


@dataclass
class Job:
    """Represents a single job execution."""

    id: str
    status: JobStatus
    reason: str | None = None
    entity_count: int | None = None
    created_at: datetime = field(default_factory=datetime.now)
    updated_at: datetime = field(default_factory=datetime.now)


class JobStore:
    """Manages jobs in memory with thread-safe operations."""

    MAX_JOBS_PER_POLICY = 5

    def __init__(self):
        """Initialize the JobStore."""
        self._lock = threading.RLock()
        self._jobs: dict[str, list[Job]] = {}  # policy_name -> jobs

    def create_job(self, policy_name: str) -> Job:
        """
        Create a new job for the given policy.

        Args:
        ----
            policy_name: Name of the policy.

        Returns:
        -------
            Job: The created job.

        """
        with self._lock:
            now = datetime.now()
            job = Job(
                id=str(uuid.uuid4()),
                status=JobStatus.RUNNING,
                created_at=now,
                updated_at=now,
            )

            # Add job to the policy's job list
            jobs = self._jobs.get(policy_name, [])
            jobs.append(job)

            # Keep only the last MAX_JOBS_PER_POLICY jobs
            if len(jobs) > self.MAX_JOBS_PER_POLICY:
                jobs = jobs[-self.MAX_JOBS_PER_POLICY :]

            self._jobs[policy_name] = jobs
            return job

    def update_job(
        self,
        policy_name: str,
        job_id: str,
        status: JobStatus,
        reason: Exception | str | None = None,
        entity_count: int | None = None,
    ) -> None:
        """
        Update the status of a job.

        Args:
        ----
            policy_name: Name of the policy.
            job_id: ID of the job to update.
            status: New status for the job.
            reason: Optional reason (exception or string).
            entity_count: Optional entity count.

        """
        with self._lock:
            jobs = self._jobs.get(policy_name, [])
            for job in jobs:
                if job.id == job_id:
                    job.status = status
                    job.updated_at = datetime.now()
                    if reason is not None:
                        if isinstance(reason, Exception):
                            job.reason = str(reason)
                        else:
                            job.reason = reason
                    if entity_count is not None:
                        job.entity_count = entity_count
                    return

    def get_jobs_for_policy(self, policy_name: str) -> list[Job]:
        """
        Get all jobs for a given policy.

        Args:
        ----
            policy_name: Name of the policy.

        Returns:
        -------
            list[Job]: List of jobs for the policy (copy to avoid race conditions).

        """
        with self._lock:
            jobs = self._jobs.get(policy_name, [])
            # Return a copy to avoid race conditions
            return [Job(**job.__dict__) for job in jobs]

    def get_all_policies_with_jobs(self) -> dict[str, list[Job]]:
        """
        Get all policies with their jobs.

        Returns
        -------
            dict[str, list[Job]]: Dictionary mapping policy names to their jobs.

        """
        with self._lock:
            result = {}
            for policy_name, jobs in self._jobs.items():
                # Return a copy to avoid race conditions
                result[policy_name] = [Job(**job.__dict__) for job in jobs]
            return result

File: network-discovery/policy/job.go
package policy

import (
	"sync"
	"time"

	"github.com/google/uuid"
)

// JobStatus represents the status of a job
type JobStatus string

const (
	// JobStatusRunning indicates the job is currently running
	JobStatusRunning JobStatus = "running"
	// JobStatusCompleted indicates the job completed successfully
	JobStatusCompleted JobStatus = "completed"
	// JobStatusFailed indicates the job failed with an error
	JobStatusFailed JobStatus = "failed"
)

// Job represents a single job execution
type Job struct {
	ID          string    `json:"id"`
	Status      JobStatus `json:"status"`
	Reason      string    `json:"reason,omitempty"`
	EntityCount int       `json:"entity_count"`
	CreatedAt   time.Time `json:"created_at"`
	UpdatedAt   time.Time `json:"updated_at"`
}

// JobStore manages jobs in memory
type JobStore struct {
	mu   sync.RWMutex
	jobs map[string][]*Job // policyName -> jobs (max 5)
}

const maxJobsPerPolicy = 5

// NewJobStore creates a new JobStore
func NewJobStore() *JobStore {
	return &JobStore{
		jobs: make(map[string][]*Job),
	}
}

// CreateJob creates a new job for the given policy and returns it
func (js *JobStore) CreateJob(policyName string) *Job {
	js.mu.Lock()
	defer js.mu.Unlock()

	now := time.Now()
	job := &Job{
		ID:        uuid.New().String(),
		Status:    JobStatusRunning,
		CreatedAt: now,
		UpdatedAt: now,
	}

	// Add job to the policy's job list
	jobs := js.jobs[policyName]
	jobs = append(jobs, job)

	// Keep only the last maxJobsPerPolicy jobs
	if len(jobs) > maxJobsPerPolicy {
		jobs = jobs[len(jobs)-maxJobsPerPolicy:]
	}

	js.jobs[policyName] = jobs
	return job
}

// UpdateJob updates the status of a job
func (js *JobStore) UpdateJob(policyName, jobID string, status JobStatus, err error, entityCount int) {
	js.mu.Lock()
	defer js.mu.Unlock()

	jobs := js.jobs[policyName]
	for _, job := range jobs {
		if job.ID == jobID {
			job.Status = status
			job.EntityCount = entityCount
			job.UpdatedAt = time.Now()
			if err != nil {
				job.Reason = err.Error()
			}
			return
		}
	}
}

// GetJobsForPolicy returns all jobs for a given policy
func (js *JobStore) GetJobsForPolicy(policyName string) []*Job {
	js.mu.RLock()
	defer js.mu.RUnlock()

	jobs := js.jobs[policyName]
	// Return a copy to avoid race conditions
	result := make([]*Job, len(jobs))
	copy(result, jobs)
	return result
}

// GetAllPoliciesWithJobs returns all policies with their jobs
func (js *JobStore) GetAllPoliciesWithJobs() map[string][]*Job {
	js.mu.RLock()
	defer js.mu.RUnlock()

	result := make(map[string][]*Job)
	for policyName, jobs := range js.jobs {
		// Return a copy to avoid race conditions
		jobsCopy := make([]*Job, len(jobs))
		copy(jobsCopy, jobs)
		result[policyName] = jobsCopy
	}
	return result
}
